jupyterlab/loader.js

```diff
@@ -0,0 +1,201 @@
+  var module = installedModules[moduleId] = {
+    exports: {},
+    id: moduleId,
+    loaded: false
+  };
+
+  // Execute the module function.
+  registered[moduleId].call(module.exports, module, module.exports, requireModule);
+
+  // Flag the module as loaded.
+  module.loaded = true;
+
+  // Return the exports of the module.
+  return module.exports;
+}
+
+
+/**
+ * Ensure a bundle is loaded on a page.
+ *
+ * @param path - The public path of the bundle (e.g. "lab/jupyter.bundle.js").
+ *
+ * @param callback - The callback invoked when the bundle has loaded.
+ */
+function ensureBundle(path, callback) {
+  // "0" is the signal for "already loaded"
+  if (installedBundles[path] === 0) {
+    return callback.call(null, requireModule);
+  }
+
+  // An array means "currently loading".
+  if (Array.isArray(installedBundles[path])) {
+    installedBundles[path].push(callback);
+    return;
+  }
+
+  // Start bundle loading.
+  installedBundles[path] = [callback];
+  var head = document.getElementsByTagName('head')[0];
+  var script = document.createElement('script');
+  script.type = 'text/javascript';
+  script.charset = 'utf-8';
+  script.async = true;
+  script.onload = function() {
+    var callbacks = installedBundles[path];
+    while(callbacks.length)
+      callbacks.shift().call(null, requireModule);
+    installedBundles[path] = 0;
+  }
+  head.appendChild(script);
+  script.src = path;
+}
+
+
+/**
+ * Ensure a list of bundles are loaded on the page.
+ *
+ * @param paths - The public paths of the bundles. (e.g. "lab/foo.bundle.js").
+ *
+ * @returns A promise that resolves with the require function when all the
+ *   bundles have loaded.
+ */
+function requireBundles(paths) {
+  var promises = paths.map(function(path) {
+    return new Promise(function (resolve, reject) {
+      ensureBundle(path, resolve);
+    });
+  });
+  return Promise.all(promises).then(function () {
+    return requireModule;
+  });
+}
+
+
+/**
+ * Find a module matching a given module request.
+ *
+ * @param name - The semver-mangled fully qualified name of the module.
+ *   For example, "foo@^1.1.0/lib/bar/baz.js".
+ *
+ * @returns The matching defined module name, if registered.  A match is
+ *   the registered name that maximally satisfies the semver range of the
+ *   request.
+ */
+function findModuleId(name) {
+  // Use the cached id if available.
+  if (lookupCache[name]) {
+    return lookupCache[name];
+  }
+  var modules = Object.keys(registered);
+  // Get the package name, semver string, and module name.
+  var parts = name.match(/(^.*?)@(.*?)(\/.*$)/) || name.match(/(^.*?)@(.*?)$/);
+  if (!parts) {
+    throw Error('Invalid module name ' + name);
+  }
+  if (parts.length === 2) {
+    parts.push('');
+  }
+  var matches = [];
+  var versions = [];
+  for (var i = 0; i < modules.length; i++) {
+    var mod = modules[i];
+    var modParts = mod.match(/(^.*?)@(.*?)(\/.*$)/) || mod.match(/(^.*?)@(.*?)$/);
```

Comment:
It would be great if we cached these parsing results at define time so that we don't have to do it for every defined module every time we look up something. Also, I think this parsing logic should be factored out into a utility function so it is easy to use/change in one place.